subsys/net/ip/connection.c

```diff
@@ -581,15 +581,25 @@ static bool conn_are_endpoints_valid(struct net_pkt *pkt, uint8_t family,
 static enum net_verdict conn_raw_socket(struct net_pkt *pkt,
 					struct net_conn *conn, uint8_t proto)
 {
-	if (proto == ETH_P_ALL) {
-		enum net_sock_type type = net_context_get_type(conn->context);
+	enum net_sock_type type = net_context_get_type(conn->context);
 
+	if (proto == ETH_P_ALL) {
 		if ((type == SOCK_DGRAM && !net_pkt_is_l2_processed(pkt)) ||
 		    (type == SOCK_RAW && net_pkt_is_l2_processed(pkt))) {
 			return NET_CONTINUE;
 		}
 	}
 
+	/*
+	 * After l2 processed only deliver protocol matched pkt,
+	 * unless the connetion protocol is all packets
+	 */
+	if (type == SOCK_DGRAM && net_pkt_is_l2_processed(pkt) &&
+	    conn->proto != htons(ETH_P_ALL) &&
+	    conn->proto != htons(net_pkt_ll_proto_type(pkt))) {
```

Comment:
Internally the network stack stores the proto in host byte order so no need to have this conversion. You can just do
```
diff --git a/subsys/net/ip/connection.c b/subsys/net/ip/connection.c
index edc0287fb0c..8025214ff16 100644
--- a/subsys/net/ip/connection.c
+++ b/subsys/net/ip/connection.c
@@ -595,8 +595,8 @@ static enum net_verdict conn_raw_socket(struct net_pkt *pkt,
         * unless the connetion protocol is all packets
         */
        if (type == SOCK_DGRAM && net_pkt_is_l2_processed(pkt) &&
-           conn->proto != htons(ETH_P_ALL) &&
-           conn->proto != htons(net_pkt_ll_proto_type(pkt))) {
+           conn->proto != ETH_P_ALL &&
+           conn->proto != net_pkt_ll_proto_type(pkt)) {
                return NET_CONTINUE;
        }
 ```
The CI tests will pass after this change.

Reply:
Thanks. Cannot access my workspace now. I'll change this later.